### PR TITLE
[Asset] Provide default context

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #19396
| License       | MIT
| Doc PR        | should be noted somewhere, ill create an issue

Allows configuring the default asset context to make things works on CLI for example. Same approach as the routing component.

Introduces
```yaml
# parameters.yml
asset.request_context.base_path: '/base/path'
asset.request_context.secure: false
```

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/assets.xml
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/assets.xml]: removing this service is a BC break, as people might reference it in their config. Please keep it

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/routing.xml
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/routing.xml]: this default is weird, as the base path is a subpart of the base url (the base url may have ``/app_dev.php`` in it)

File: src/Symfony/Component/Asset/Context/ContextFactory.php
Comment (stof) [inline src/Symfony/Component/Asset/Context/ContextFactory.php]: this is wrong, as the request may be injected in the request stack later than that. You cannot instantiate the service based on the availability of the Request, as it would change the behavior depending on when the context is instantiated (which depends on what needs it in the container and so is unknown to you)

File: src/Symfony/Component/Asset/Context/RequestContextContext.php
Comment (stof) [inline src/Symfony/Component/Asset/Context/RequestContextContext.php]: should be detected based on the RequestContext scheme instead

File: src/Symfony/Component/Routing/RequestContext.php
Comment (stof) [inline src/Symfony/Component/Routing/RequestContext.php]: this looks weird to me, because the Routing component never needs the base path

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/assets.xml
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/assets.xml]: Can we deprecate it?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/routing.xml
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/routing.xml]: Agree.. `base_url` should default to `base_path` agree?

File: src/Symfony/Component/Asset/Context/ContextFactory.php
Comment (ro0NL) [inline src/Symfony/Component/Asset/Context/ContextFactory.php]: Correct. What about a direct `PHP_SAPI` check?

File: src/Symfony/Component/Routing/RequestContext.php
Comment (ro0NL) [inline src/Symfony/Component/Routing/RequestContext.php]: True, however for being a _request_ context it seemed a reasonable feature to me. And more important, allows this fix :))

**Comment by javiereguiluz** [comment]
@ro0NL I recommend you to stop working on this feature. The reason is that the related issue hasn't been approved at all as a legit new feature. (Personally I'm not sure it's needed at all). It's always better to work only on approved features and verified bugs. Thanks!

**Comment by ro0NL** [comment]
Imo. the behavior is buggy, and this is my approach to revise it (i may do PR's rigth?).

I explained why imo. it's needed, so feel free to argue.

**Comment by ro0NL** [comment]
There are 2 viable options to me;
- this approach 
- introduce a new `%asset.default_base_path%` param and provide a default context

Option is also reasonable, and may perhaps be the better approach. However i chose this one as we already introduced routing params, and it will work for people who already set the base url with it.

**Comment by ro0NL** [comment]
@stof different approach, way simpler :) WDYT?

**Comment by nicolas-grekas** [comment]
Since this adds new parameters, this is not a bug but only a feature.
Doc PR should be opened before merging, because that's the only way to know about the params.
About the params, they are obscure things usually to users, which means very little are going to know about them and their purpose, even if it's documented.
Which means we shouls try hard to *not* add new params when possible (I don't know here, just emphasizing this important point :) .)

**Comment by ro0NL** [comment]
Agree.. on the params. Im not even sure we need them, same for routing context params actually; see #21043

However the bug is real.

**Comment by xabbuh** [comment]
Before the introduction of the new Asset component the `asset()` function made use of the request context config options, right?

**Comment by ro0NL** [comment]
Not sure, i should have a look at that. Fact is (given the current context options) we cannot go from base url to base path genericly.

```
/request-base/app_dev.php/route
/request-base/asset-base/asset vs. /request-base/app_dev.php/asset-base/asset
```

**Comment by ro0NL** [comment]
I guess practically one could get away with configuring package base urls, going absolute. Bypassing `RequestStackContext` and no SSL support :)

**Comment by xabbuh** [comment]
As this used to work in previous releases we should at least strive to make the Asset component compatible with how the Twig functions used to work before. We can still think about how to make things better and deprecate stuff in 3.3/3.4.

**Comment by ro0NL** [comment]
> As this used to work in previous releases we should at least strive to make the Asset component compatible with how the Twig functions used to work before.

Im not sure i follow.. i did a quik test; installed latest 2.7 in a subdir (reachable at `http://symfony-sub.dev/sf27/web/app_dev.php`)

Updated framework config with `assets: { version: 1, base_path: /assets }`

Updated `app/Resources/views/default/index.html.twig` to display `{{ asset('foo.jpg') }}`

Added a command `test` with `$output->writeln($this->getContainer()->get('twig')->render(':default:index.html.twig'));`

## Results :)

image

image

What did we removed? Or what am i missing here?

**Comment by xabbuh** [comment]
I mean before the introduction of the Asset component (so you will have to use a Symfony version before 2.7 to see how it used to behave back then).

**Comment by ro0NL** [comment]
I see. Not gonna investigate that now :) I mean, i cannot really compare the asset component against nothing. Either the approach was the same-ish, or it suffered from the same problem. Here's my approach for 2.7 and up :+1: (perhaps along with #21043).

But again, if im missing something in terms of a different approach done previously.. let me know :)

**Comment by ro0NL** [comment]
Oh, and it's totally fixable in user land. So we can always keep things as is here.

**Comment by xabbuh** [comment]
> Oh, and it's totally fixable in user land. So we can always keep things as is here; an unsupported feature.

But the thing is that this used to work before (the `asset()` function did exist before the introduction of the Asset component in #13234, see https://github.com/symfony/symfony/blob/2.3/src/Symfony/Bundle/TwigBundle/Extension/AssetsExtension.php#L55).

**Comment by ro0NL** [comment]
I think looking at https://github.com/symfony/symfony/blob/2.3/src/Symfony/Bundle/FrameworkBundle/Templating/Asset/PathPackage.php#L31 and https://github.com/symfony/symfony/blob/2.3/src/Symfony/Component/Templating/Asset/PathPackage.php#L35 it suffered the same problem.

But im really not aware of any CLI defaulting with params or so in 2.3, which would basically be this approach.

**Comment by ro0NL** [comment]
Now we could do it convention wise perhaps, making use of `%router.request_context.base_url%`.

If ending with `/` it's the base path (`/request-base/`), otherwise it's the base URL (`/request-base/app_dev.php`) and we do a `dirname`. Could work...

This would keep a single param for prod envs.

**Comment by xabbuh** [comment]
@ro0NL I'll look into an old 2.3 application to check how exactly it used to work then.

**Comment by ro0NL** [comment]
@xabbuh any news? :P

Im planning to close; as it's solveable in userland i guess by using a packge base URL (coming from a custom param or so).

Prettiest fix is still #21043 IMO. But this PR should do as well :-)

Otherwise im planning to close; as it's solvable in userland by using a package base URL (coming from a custom param or so).

> About the params, they are obscure things usually to users, which means very little are going to know about them and their purpose, even if it's documented.

I tend to believe people will bump into it (if documented somewhere of course) once dealing with this issue =/

**Comment by nicolas-grekas** [comment]
can you add a changelog line to advertise the feature?
also, should we add corresponding config options? not sure, just wondering.

**Comment by ro0NL** [comment]
Changelog added; extra config is not needed IMHO; it's the same approach as routing does.

Because of that ive renamed the params to `%asset.request_context.*%` so it's in line with `%router.request_context.*%`

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: we could add an entry to the changelog file of the Asset component too

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: and hint a bit more why they're useful :)

**Comment by ro0NL** [comment]
changelog updated. Not sure about a note in assets component; change is in no way related to that.

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
ping @fabpot as I guess you're the one merging on Asset?

**Comment by ro0NL** [comment]
Also im not experiencing this issue myself. Honestly thought to fix _linked issue_ real quick :) confirmed as it is (https://github.com/symfony/symfony/issues/19396#issuecomment-234212063).

Not sure what's blocking here. However i dont wanna imply im pushing this PR or so.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ro0NL.

**Pull request closed without merging**
